sigstore/_internal/trustroot.py

```diff
@@ -102,72 +106,119 @@ class KeyringSignatureError(KeyringError):
     """
 
 
-class Keyring:
+@dataclass(init=False)
+class Key:
     """
-    Represents a set of CT signing keys, each of which is a potentially
-    valid signer for a Signed Certificate Timestamp (SCT).
-
-    This structure exists to facilitate key rotation in a CT log.
+    Represents a key in a `Keyring`.
     """
 
-    def __init__(self, keys: List[bytes] = []):
+    hash_algorithm: hashes.HashAlgorithm
+    key: PublicKey
+    key_id: KeyID
+
+    _RSA_SHA_256_DETAILS: ClassVar[set[_PublicKeyDetails]] = {
+        _PublicKeyDetails.PKCS1_RSA_PKCS1V5,
+        _PublicKeyDetails.PKIX_RSA_PKCS1V15_2048_SHA256,
+        _PublicKeyDetails.PKIX_RSA_PKCS1V15_3072_SHA256,
+        _PublicKeyDetails.PKIX_RSA_PKCS1V15_4096_SHA256,
+    }
+
+    _EC_DETAILS_TO_HASH: ClassVar[dict[_PublicKeyDetails, hashes.HashAlgorithm]] = {
+        _PublicKeyDetails.PKIX_ECDSA_P256_SHA_256: hashes.SHA256(),
+        _PublicKeyDetails.PKIX_ECDSA_P384_SHA_384: hashes.SHA384(),
+        _PublicKeyDetails.PKIX_ECDSA_P521_SHA_512: hashes.SHA512(),
+    }
+
```

Comment:
no need to do this here, but is it relatively straightforward to add ed25519 down the line?

Reply:
Yep! It should just be another branch (below) in the future.